raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -215,34 +215,53 @@ private void applyCommittedRecordsToStateMachine() {
         quorum.highWatermark().ifPresent(highWatermark -> {
             log.updateHighWatermark(highWatermark);
             maybeCommitPendingAppends(highWatermark, time.milliseconds());
+        });
+    }
 
-            logger.debug("Applying committed entries up to high watermark {}. " +
+    private void maybeCommitPendingAppends(long highWatermark, long currentTimeMs) {
+        logger.debug("Applying committed entries up to high watermark {}. " +
                 "Current position is {}", highWatermark, stateMachine.position());
 
-            while (stateMachine.position().offset < highWatermark && shutdown.get() == null) {
-                OffsetAndEpoch position = stateMachine.position();
+        Iterator<Map.Entry<OffsetAndEpoch, AppendedInfo>> iter = uncommittedAppends.entrySet().iterator();
+        while (stateMachine.position().offset < highWatermark && shutdown.get() == null) {
+            OffsetAndEpoch position = stateMachine.position();
+
+            if (iter.hasNext()) {
+                Map.Entry<OffsetAndEpoch, AppendedInfo> entry = iter.next();
+                AppendedInfo appendedInfo = entry.getValue();
+
+                if (appendedInfo.baseOffset < position.offset) {
+                    iter.remove();
+                } else if (appendedInfo.baseOffset > position.offset) {
+                    // the cached uncommitted appends may be lost when, e.g. the client crashed;
+                    // in this case we'd still have to read from the local log
+                    Records records = readCommitted(position, highWatermark);
+
+                    // TODO: when we have log compaction, this logic needs to be updated
+                    stateMachine.apply(records, 0);
+                    logger.trace("Applied committed records at {} to the state machine; position " +
+                            "updated to {}", position, stateMachine.position());
+                } else if (appendedInfo.lastOffset < highWatermark) {
+                    // note we assume that the high-watermark would never advances partial batches, but would always
+                    // on batch boundaries; if it is not the case then this logic would break
+                    double elapsedTimePerRecord = (currentTimeMs - appendedInfo.appendTimeMs) / (double) appendedInfo.numRecords;
+                    kafkaRaftMetrics.updateCommitLatency(elapsedTimePerRecord, currentTimeMs);
+
+                    Records records = appendedInfo.records;
+                    long baseOffset = appendedInfo.baseOffset - records.records().iterator().next().offset();
```

Comment:
This is not elegant, I'm just hacking it around to make simulation pass while illustrating the idea.

The baseOffset would be non-zero for leader-appends, and it would be zero for follower-appends.